aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu

```diff
@@ -0,0 +1,472 @@
+#include "ATen/ATen.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/Utils.h"
+#include "c10/util/Exception.h"
+#include <THC/THCGeneral.h>
+#include "THC/THCNumerics.cuh"
+
+#include <algorithm>
+#include <cfloat>
+#include <cmath>
+
+#define CUDA_MAX_THREADS 1024   // this is safe, in reality 256 is our limit
+
+#define START_IND(a,b,c) (int)std::floor((float)(a * c) / b)
+#define END_IND(a,b,c) (int)std::ceil((float)((a + 1) * c) / b)
+// #define START_IND(a,b,c) a * c / b
+// #define END_IND(a,b,c)  (a + 1) * c / b + ((a + 1) * c % b > 0)?1:0
+
+
+namespace at {
+namespace native {
+
+namespace {
+
+// 4d tensor B x D x H x W
+
+/*
+ * Description:
+ *    this function adaptively maxpools an input 4D tensor along dimensions 2 and 3
+ *    4D input, 4D output, 4D argmax x and y
+ */
+ template <typename T>
```

Comment:
using `T` here would be fine as it is also used in https://github.com/pytorch/pytorch/pull/14714/files#diff-7b64cf245203cfc8f04802bd74787113R36 .. but maybe would be good to change that to `scalar_t` as used in other files such as https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L29

Reply:
Yeah, I don't think it's a big deal either way, people can figure it out.